File: app/src/main/java/net/xzos/UpgradeAll/utils/VersionChecker.java
package net.xzos.UpgradeAll.utils;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import androidx.annotation.NonNull;

import com.jaredrummler.android.shell.CommandResult;
import com.jaredrummler.android.shell.Shell;

import net.xzos.UpgradeAll.application.MyApplication;
import net.xzos.UpgradeAll.server.log.LogUtil;

import org.apache.maven.artifact.versioning.DefaultArtifactVersion;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class VersionChecker {

    private static final String TAG = "VersionChecker";
    private static final String[] LogObjectTag = {"Core", TAG};
    protected static final LogUtil Log = MyApplication.getServerContainer().getLog();

    private JSONObject versionCheckerJsonObject;

    public VersionChecker(JSONObject versionCheckerJsonObject) {
        this.versionCheckerJsonObject = versionCheckerJsonObject;
    }

    public String getVersion() {
        JSONObject versionCheckerJsonObject = this.versionCheckerJsonObject;
        String versionCheckerApi = null;
        String version = null;
        try {
            versionCheckerApi = versionCheckerJsonObject.getString("api");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        if (versionCheckerApi != null) {
            String shellCommand;
            try {
                shellCommand = versionCheckerJsonObject.getString("text");
            } catch (JSONException e) {
                Log.e(LogObjectTag, TAG, "getVersion: JSONObject解析出错,  versionCheckerJsonObject: " + versionCheckerJsonObject.toString());
                shellCommand = "";
            }
            switch (versionCheckerApi.toLowerCase()) {
                case "app":
                    version = getAppVersion(versionCheckerJsonObject);
                    break;
                case "magisk":
                    version = getMagiskModuleVersion(versionCheckerJsonObject);
                    break;
                case "shell":
                    version = Shell.run(shellCommand).getStdout();
                    break;
                case "shell_root":
                    version = Shell.SU.run(shellCommand).getStdout();
                    break;
            }
        }
        return version;
    }

    private String getAppVersion(@NonNull JSONObject versionCheckerJsonObject) {
        // 获取软件版本
        String appVersion = null;
        String packageName = null;
        try {
            packageName = versionCheckerJsonObject.getString("text");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        if (packageName != null) {
            Context context = MyApplication.getContext();
            try {
                PackageInfo packageInfo = context.getPackageManager().getPackageInfo(packageName, 0);
                appVersion = packageInfo.versionName;
            } catch (PackageManager.NameNotFoundException e) {
                appVersion = null;
            }
        }
        return appVersion;
    }

    private String getMagiskModuleVersion(@NonNull JSONObject versionCheckerJsonObject) {
        String magiskModuleVersion = null;
        String magiskModuleName = null;
        try {
            magiskModuleName = versionCheckerJsonObject.getString("text");
        } catch (JSONException e) {
            e.printStackTrace();
        }
        String modulePropFilePath = "/data/adb/modules/" + magiskModuleName + "/module.prop";
        String command = "cat " + modulePropFilePath;
        CommandResult result = Shell.SU.run(command);
        String[] resultList = result.getStdout().split("\n");
        String keyWords = "version=";
        for (String resultLine : resultList) {
            if (resultLine.indexOf(keyWords) == 0) {
                magiskModuleVersion = resultLine.substring(keyWords.length());
            }
        }
        return magiskModuleVersion;
    }

    private static boolean isVersion(String versionString) {
        boolean isVersion = false;
        String regexString = "\\d+(\\.\\d+)*";
        if (versionString != null) {
            Pattern p = Pattern.compile(regexString);
            Matcher m = p.matcher(versionString);
            if (m.find()) {
                isVersion = true;
            }
        }
        return isVersion;
    }

    private static String getVersionNumberString(@NonNull String versionString) {
        String versionMatchString = versionString;
        String[] versionStringList = versionString.split(" ");
        for (String temp : versionStringList) {
            if (isVersion(temp))
                versionMatchString = temp;
        }
        return versionMatchString;
    }

    public static boolean compareVersionNumber(String versionNumber0, String versionNumber1) {
        /*
         * 对比 versionNumber0 与 versionNumber1
         * 若，前者比后者大，则返回 true*/
        Log.i(LogObjectTag, TAG, String.format("compareVersionNumber: versionNumber0: %s , versionNumber1: %s", versionNumber0, versionNumber1));
        versionNumber0 = getVersionNumberString(versionNumber0);
        versionNumber1 = getVersionNumberString(versionNumber1);
        DefaultArtifactVersion version0 = new DefaultArtifactVersion(versionNumber0);
        DefaultArtifactVersion version1 = new DefaultArtifactVersion(versionNumber1);
        return version0.compareTo(version1) >= 0;
    }
}

File: app/src/main/kotlin/net/xzos/upgradeAll/utils/VersionChecker.kt
package net.xzos.upgradeAll.utils

import android.annotation.SuppressLint
import com.google.gson.Gson
import com.jaredrummler.android.shell.Shell
import net.xzos.upgradeAll.R
import net.xzos.upgradeAll.application.MyApplication
import net.xzos.upgradeAll.data.json.gson.AppConfig
import net.xzos.upgradeAll.data.json.gson.VersionCheckerGson
import net.xzos.upgradeAll.server.ServerContainer
import org.apache.maven.artifact.versioning.DefaultArtifactVersion
import java.util.regex.Pattern

class VersionChecker(private val versionCheckerGson: AppConfig.AppConfigBean.TargetCheckerBean?) {

    val version: String?
        get() {
            val versionCheckerApi: String? = versionCheckerGson?.api
            var version: String? = null
            if (versionCheckerApi != null) {
                val shellCommand: String? = versionCheckerGson?.extraString

                if (shellCommand != null)
                    @SuppressLint("DefaultLocale")
                    when (versionCheckerApi.toLowerCase()) {
                        "app_package" -> version = getAppVersion()
                        "magisk_module" -> version = getMagiskModuleVersion()
                        "shell" -> version = Shell.run(shellCommand).getStdout()
                        "shell_root" -> version = Shell.SU.run(shellCommand).getStdout()
                    }
            }
            return version
        }

    private fun getAppVersion(): String? {
        // 获取软件版本
        return try {
            val packageInfo = MyApplication.context.packageManager.getPackageInfo(versionCheckerGson?.extraString!!, 0)
            packageInfo.versionName
        } catch (e: Throwable) {
            null
        }
    }

    private fun getMagiskModuleVersion(): String? {
        var magiskModuleVersion: String? = null
        val modulePropFilePath = "/data/adb/modules/${versionCheckerGson?.extraString}/module.prop"
        val command = "cat $modulePropFilePath"
        val result = Shell.SU.run(command)
        val resultList = result.getStdout().split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val keyWords = "version="
        for (resultLine in resultList) {
            if (resultLine.indexOf(keyWords) == 0) {
                magiskModuleVersion = resultLine.substring(keyWords.length)
            }
        }
        return magiskModuleVersion
    }

    companion object {

        private const val TAG = "VersionCheckerGson"
        private val LogObjectTag = arrayOf("Core", TAG)
        private val Log = ServerContainer.Log

        private fun getVersionNumberString(versionString: String?): String? {
            var versionMatchString: String? = null
            val regexString = MyApplication.context.getString(R.string.versioning_regex_match)
            if (versionString != null) {
                val p = Pattern.compile(regexString)
                val m = p.matcher(versionString)
                versionMatchString = if (m.find()) {
                    m.group()
                } else
                    versionString
            }
            return versionMatchString
        }

        /**
         * 对比 versionNumber0 与 versionNumber1
         * 若，前者比后者大，则返回 true*/
        fun compareVersionNumber(versionNumber0: String?, versionNumber1: String?): Boolean {
            @Suppress("NAME_SHADOWING")
            var versionNumber0 = versionNumber0
            @Suppress("NAME_SHADOWING")
            var versionNumber1 = versionNumber1
            Log.i(LogObjectTag, TAG, String.format("compareVersionNumber: versionNumber0: %s, versionNumber1: %s", versionNumber0, versionNumber1))
            versionNumber0 = getVersionNumberString(versionNumber0)
            versionNumber1 = getVersionNumberString(versionNumber1)
            if (versionNumber0 != null && versionNumber1 != null) {
                val version0 = DefaultArtifactVersion(versionNumber0)
                val version1 = DefaultArtifactVersion(versionNumber1)
                return version0 >= version1
            }
            return false
        }

        /** 修补老标准格式
         * TODO: 修改版本: 0.1.0-alpha.beta
         */
        @SuppressLint("DefaultLocale")
        fun fixJson(jsonString: String): AppConfig.AppConfigBean.TargetCheckerBean {
            return AppConfig.AppConfigBean.TargetCheckerBean().apply {
                val versionCheckerGson = Gson().fromJson(jsonString, VersionCheckerGson::class.java)
                val extraString = versionCheckerGson.text
                val versionCheckerApi = versionCheckerGson.api
                if (extraString != null && versionCheckerApi != null) {
                    this.extraString = versionCheckerGson.text
                    this.api = when (versionCheckerApi.toLowerCase()) {
                        "app" -> "App_Package"
                        "magisk" -> "Magisk_Module"
                        else -> versionCheckerApi
                    }
                }
            }
        }
    }
}